api/auth.go

```diff
@@ -4,49 +4,44 @@ import (
 	"context"
 	"fmt"
 	"net/http"
-	"time"
 
 	"github.com/Sirupsen/logrus"
 	jwt "github.com/dgrijalva/jwt-go"
 	"github.com/netlify/gocommerce/claims"
 	gcontext "github.com/netlify/gocommerce/context"
+	"github.com/netlify/gocommerce/models"
 )
 
-func (a *API) withToken(ctx context.Context, w http.ResponseWriter, r *http.Request) context.Context {
-	log := gcontext.GetLogger(ctx)
+func withTokenCtx(w http.ResponseWriter, r *http.Request) (context.Context, error) {
+	ctx := r.Context()
+	log := getLogEntry(r)
 	config := gcontext.GetConfig(ctx)
 	authHeader := r.Header.Get("Authorization")
 	if authHeader == "" {
 		log.Info("Making unauthenticated request")
-		return ctx
+		return ctx, nil
 	}
 
 	matches := bearerRegexp.FindStringSubmatch(authHeader)
 	if len(matches) != 2 {
-		log.Info("Invalid auth header format: " + authHeader)
-		unauthorizedError(w, "Bad authentication header")
-		return nil
+		return nil, unauthorizedError("Bad authentication header").WithInternalMessage("Invalid auth header format: %s", authHeader)
 	}
 
 	token, err := jwt.ParseWithClaims(matches[1], &claims.JWTClaims{}, func(token *jwt.Token) (interface{}, error) {
-		if token.Header["alg"] != jwt.SigningMethodHS256.Name {
-			return nil, fmt.Errorf("Unexpected signing method: %v", token.Header["alg"])
+		if token.Method.Alg() != jwt.SigningMethodHS256.Name {
+			return nil, fmt.Errorf("Unexpected signing method: %v", token.Method.Alg())
```

Comment:
hmmmm that is a 401 is fine. I think being explicit would be nice though, just for clarity though it may be redundant.